### PR TITLE
[Music] Validator for sounds fields

### PR DESCRIPTION
**Pull request by mikeharv**
If a level with start sources has it's selected library updated, it can invalidate any field values in sound blocks:

image
image

Blockly supports field validators as a means to make sure that fields always end up with a valid value. 
https://developers.google.com/blockly/guides/create-custom-blocks/fields/validators

We can implement a field validator that will check that the sound id can be found in the current MusicLibrary instance. If it's not, the validator can return the default sound (which matches the value we use in the toolbox). 

image

This came up because we realized that a whole bunch of new levels were not using the recommended `launch2024` library yet, and switching it was breaking blocks. 
In practice, a levelbuilder may still need to revise their start sources after changing libraries, but this will prevent invalid values from being passed to the sequencer. The alternative that was considered was that switching libraries could trigger a hard reset of start sources, which feels much more destructive.

### CODE REVIEW

**Comment by breville** [comment]
Does this mean that after changing library, and reentering the Blockly-based start sources editor (rather than the JSON one) then the blocks will all have their values reset, and can then be updated?

**Comment by mikeharv** [comment]
> Does this mean that after changing library, and reentering the Blockly-based start sources editor (rather than the JSON one) then the blocks will all have their values reset, and can then be updated?

@breville Yes, good question! This doesn't actually change the start sources JSON. It just prevents invalid values from being used with Blockly. If the start sources were not updated afterwards, this would also prevent a student from seeing a sound field without a valid value.

**Comment by breville** [comment]
> > Does this mean that after changing library, and reentering the Blockly-based start sources editor (rather than the JSON one) then the blocks will all have their values reset, and can then be updated?
> 
> @breville Yes, good question! This doesn't actually change the start sources JSON. It just prevents invalid values from being used with Blockly. If the start sources were not updated afterwards, this would also prevent a student from seeing a sound field without a valid value.

Understood.  Hopefully the levelbuilder flow will more usually be to edit start sources after changing library so that we do the fix-up before saving the level, rather than at user-facing time, though this code will handle things if invalid sound names make it to there.

**Review by breville — APPROVED** [review]

File: apps/src/music/blockly/extensions.js [hunk 1]
```diff
@@ -151,3 +153,13 @@ export const effectsFieldExtension = function () {
     baseHandler.call(fieldEffectsName, menu, menuItem);
   };
 };
+
+export const fieldSoundsValidator = function () {
+  this.getField('sound').setValidator(newValue => {
```
Comment (sanchitmalhotra126) [inline apps/src/music/blockly/extensions.js]: minor: can use the `FIELD_SOUNDS_NAME` constant here to keep it consistent with the field definition

**Review by sanchitmalhotra126 — APPROVED** [review]
Very cool! I wonder if this will drive down some of the similar sound warnings we've been seeing in our logs.

File: apps/src/music/blockly/extensions.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/music/blockly/extensions.js]: Oh good call! I meant to check to see if I could find a constant for this. Thanks for helping me find it.

**Pull request merged**